### PR TITLE
Add bool flags to argument parser

### PR DESCRIPTION
**Pull request by arezaii**
This PR adds boolean flags to the argument parser. 

Scope of work defined in Cray/chapel-private#2301, this work satisfies
the main goals of that issue which are to:

1. Allow the developer to add boolean flags to their CLI
2. Allow options, subcommands, and flags to be entered together

Relates to broader implementation task in Cray/chapel-private#2192

Tests added for flags and mixing subcommands with options and flags.

Example updated to include flags.

TESTING:

After running `make cleanall` from `$CHPL_HOME`:

- [x] Can `make`
- [x] Can `make docs`
- [x] Can `make mason`
- [x] Can `make check`
- [x] Passing tests from `util/start_test test/mason/mason-argparse`

Reviewed by @mppf 

Signed-off-by: Ahmad Rezaii ahmad.rezaii@hpe.com

### CODE REVIEW

**Review by mppf — APPROVED** [review]

File: tools/mason/MasonArgParse.chpl [hunk 1]
```diff
@@ -132,6 +147,92 @@ module MasonArgParse {
...
+    override proc _isRequired() {
+      return this._required;
+    }
+
+    override proc _validate(present:bool, valueCount:int):string {
+      debugTrace("expected " + _numArgs:string + " got " + valueCount:string);
+      debugTrace("present="+present:string + " required="+_required:string);
+      if !present && _required {
+        return "Required value missing";
+      } else if valueCount < _numArgs.low && present {
```
[the remark this answers is not in the document]
Comment (arezaii) [reply tools/mason/MasonArgParse.chpl:194]: In most cases, it's not necessary. However, in the [discussion about flags](https://github.com/Cray/chapel-private/issues/2192#issuecomment-878719883), @bradcray suggested the following:

> I'm proposing that if a boolean argument expects 0 arguments, it must be written as -f; that if it accepts 1, it must be written as -f=true or -f true; and if it accepts 0..1 arguments, any of these forms can be used.

In that case, the valueCount is used to determine if the flag (-f) is followed by a value (true/false) or not. If the developer specified the lowerBound is 1, then valueCount must be 1 or else we error.

File: tools/mason/MasonArgParse.chpl [hunk 2]
```diff
@@ -357,34 +458,104 @@ module MasonArgParse {
...
+                     opts=opts,
+                     required=required,
+                     defaultValue=defaultValue,
+                     flagInversion=flagInversion,
+                     numArgs=numArgs..numArgs);
+    }
+
+    proc addFlag(name:string, opts:[?optsD],
+                 required=false, defaultValue:?t=none, flagInversion=true,
+                 numArgs:range) throws {
```
[the remark this answers is not in the document]
Comment (arezaii) [reply tools/mason/MasonArgParse.chpl:497]: Both overloads should be available to the developer as they may wish to force the user to enter a fixed number of values after the flag `-f` or specify that 0..1 values are acceptable, allowing the user to enter any of the following `-f` or `-f=false` or `-f false`.

File: tools/mason/MasonArgParse.chpl [hunk 3]
```diff
@@ -454,7 +624,19 @@ module MasonArgParse {
       return _addAction(action);
     }
   }
-  
+
+  // helper to convert string values to booleans
+  proc _convertStringToBool(strVal:string, inout boolVal:bool) : bool {
+    if strVal.strip(" ") == "1" || strVal.toLower() == "true" {
```
[the remark this answers is not in the document]
Comment (arezaii) [reply tools/mason/MasonArgParse.chpl:630]: agree on all points

File: tools/mason/MasonArgParse.chpl [hunk 4]
```diff
@@ -132,6 +147,92 @@ module MasonArgParse {
...
+        return "Not enough values: expected " + _numArgs:string +
+               " got " + valueCount:string;
+      } else {
+        return "";
+      }
+    }
+
+    override proc _match(args:[?argsD]string, startPos:int, myArg:Argument,
+                         ref rest:list(string), endPos:int) throws {
+      var high = _numArgs.high;
```
[the remark this answers is not in the document]
Comment (arezaii) [reply tools/mason/MasonArgParse.chpl:204]: This same parameter name is used for each type of option/argument/flag. I didn't think it was too strange as Python does this in a similar fashion, where their `nargs` parameter can specify a single integer count or several allowable counts via the `+`, `?`, or `*` values. However, we can certainly update it everywhere when we identify a more compelling/meaningful name.

File: tools/mason/MasonArgParse.chpl (hunk 2 above)
[the remark this answers is not in the document]
Comment (mppf) [reply tools/mason/MasonArgParse.chpl:497]: Does it make sense to have a fixed number of arguments after a flag? Apart from `-f false` it seems to me a flag shouldn't allow any number of arguments and that should be handled by a different sort of Argument.

File: tools/mason/MasonArgParse.chpl (hunk 4 above)
[the remark this answers is not in the document]
Comment (mppf) [reply tools/mason/MasonArgParse.chpl:204]: Yeah, it's making more sense now that I am understanding more about how it is a generalization of an integer.

File: tools/mason/MasonArgParse.chpl (hunk 2 above)
[the remark this answers is not in the document]
Comment (arezaii) [reply tools/mason/MasonArgParse.chpl:497]: If a developer desires more than 1 value after a flag then they should use an option. Flags expect only 0, 1, or 0..1 values. It is conceivable, although not recommended, that a developer would want to force the user to enter a true/false value for the flag, and the way they would do that is by specifying that the flag expects 1 value to follow.

**Pull request merged**